File: src/engine_js/matching.js
/**
 * matching.js
 * -----------
 * Port of src/engine/matching.py -- no behavior change.
 */

/**
 * @param {import('./models').CardRequirement[]} requirements
 * @param {Map<string, object>} collectionIndex
 * @returns {Array<object>} per-card match records (plain objects, same
 *   shape as the Python dict version -- not elevated to a public type,
 *   see models.js docstring for why)
 */
function matchDeck(requirements, collectionIndex) {
  const results = [];

  for (const req of requirements) {
    const required = req.quantity;
    const collection = collectionIndex.get(req.name);

    let ownedTotal = 0;
    if (collection) {
      const n = Number(collection.total_quantity);
      ownedTotal = Number.isFinite(n) ? n : 0;
    }

    const owned = Math.min(Math.max(ownedTotal, 0), required);
    const missing = required - owned;

    results.push({
      name: req.name,
      quantity: req.quantity,
      finish: req.finish,
      scryfallId: req.scryfallId,
      board: req.board,
      required,
      owned,
      missing,
      collectionRecord: collection ?? null,
    });
  }

  return results;
}

module.exports = { matchDeck };

File: src/engine/matching.py
"""
Matching requirements against the owned collection.

Extracted from deck_cost.py's match_deck() with no behavior change.
The per-card match record stays a plain dict on purpose (see models.py
docstring) -- only the deck-level MatchResult is a public dataclass,
and engine.py builds that by summing over these dicts.
"""

from __future__ import annotations

from typing import Any

from .models import CardRequirement


def match_deck(
    requirements: list[CardRequirement],
    collection_index: dict[str, dict[str, Any]],
) -> list[dict[str, Any]]:
    results = []

    for req in requirements:
        required = req.quantity
        collection = collection_index.get(req.name)

        owned_total = 0
        if collection:
            try:
                owned_total = int(collection.get("total_quantity", 0))
            except (TypeError, ValueError):
                owned_total = 0

        owned = min(max(owned_total, 0), required)
        missing = required - owned

        results.append({
            "name": req.name,
            "quantity": req.quantity,
            "finish": req.finish,
            "scryfall_id": req.scryfall_id,
            "board": req.board,
            "required": required,
            "owned": owned,
            "missing": missing,
            "collection_record": collection,
        })

    return results
